src/coreclr/jit/optimizer.cpp

```diff
@@ -9657,9 +9701,15 @@ GenTree* OptBoolsDsc::optIsBoolComp(OptTestInfo* pOptTest)
     assert(pOptTest->testTree->gtOper == GT_JTRUE || pOptTest->testTree->gtOper == GT_RETURN);
     GenTree* cond = pOptTest->testTree->AsOp()->gtOp1;
 
-    // The condition must be "!= 0" or "== 0"
+    // The condition must be "!= 0" or "== 0" or >=0 or <0
+
+    if (!cond->OperIs(GT_EQ, GT_NE, GT_LT, GT_GE))
+    {
+        return nullptr;
+    }
 
-    if ((cond->gtOper != GT_EQ) && (cond->gtOper != GT_NE))
+    // we don't optimize this statements because we might delete them (e.g. array range checks)
+    if (cond->OperIs(GT_LT, GT_GE) && (m_b1->bbFlags != BBF_IMPORTED || m_b2->bbFlags != BBF_IMPORTED))
```

Comment:
This flag check looks strange to me.

Reply:
array range checks instructions (indexer >= 0, indexer < array.Length) are added to the user code, but later they get deleted during the Range Check Elimination step. When those instructions are passed through the boolean optimization step they get optimized with user code, so they're no longer deleted. I think @AndyAyersMS has done some fixes so I don't need it anymore, I'll delete it and see the results

Reply:
things worked out for the better in general, had light regressions here :
* libraries_tests.pmi.OSX.arm64.checked.mch : had 8 diffs, now 132
* libraries.pmi.windows.arm64.checked.mch : had 264 diffs, now 484
* aspnet.run.windows.x64.checked.mch : had -37 diffs, now 93
* We improved the diffs in all the other configurations

Reply:
The check seems to be quite opportunistic and works by change -- it can introduce spurious diffs in the future if any unrelated BB flags are added. We should try to dial into some specific properties we can check to accomplish this. I'll set some time aside and look into it next week.

Reply:
Ah, I realized you meant that you removed the check. Ok, then the change looks fine for me, though I do see some large percent wise regressions:
```scala
Top method regressions (percentages):
         250 (41.88 % of base) : 139400.dasm - System.Speech.Internal.StringBlob:.ctor
          58 (33.33 % of base) : 149739.dasm - Microsoft.Extensions.Primitives.StringSegment:IndexOf(ushort,int,int):int:this
          34 (31.48 % of base) : 149718.dasm - Microsoft.Extensions.Primitives.StringSegment:AsSpan(int,int):System.ReadOnlySpan`1[ushort]:this
         154 (30.02 % of base) : 211775.dasm - System.Net.HttpWebRequest:AddRange(System.String,long,long):this
          25 (15.43 % of base) : 208543.dasm - System.Net.Http.Headers.HeaderUtilities:TryParseInt32(System.String,int,int,byref):bool
          25 (15.34 % of base) : 208544.dasm - System.Net.Http.Headers.HeaderUtilities:TryParseInt64(System.String,int,int,byref):bool
          22 (13.17 % of base) : 149740.dasm - Microsoft.Extensions.Primitives.StringSegment:IndexOf(ushort,int):int:this
          27 (10.67 % of base) : 208840.dasm - System.Net.Http.Headers.Int32NumberHeaderParser:GetParsedValueLength(System.String,int,System.Object,byref):int:this
```
@AndyAyersMS I'm not sure what the status ended up being here. It seems like the phase was meant to be moved later from some of your latest comments, but it still seems to be running at the same point and this looks like some of the same regressions you talked about above.

Reply:
(FWIW, all the force pushing that has been done in this PR makes it very hard to see what happened. In the future it would be nice to avoid force pushing.)